modules/python/python-http-header.c

```diff
@@ -0,0 +1,390 @@
+/*
+ * Copyright (c) 2020 One Identity
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "python-module.h"
+#include "python-helpers.h"
+
+#include "lib/driver.h"
+#include "lib/str-utils.h"
+#include "lib/list-adt.h"
+#include "modules/http/http-signals.h"
+
+#include "python-http-header.h"
+
+#include <time.h>
+
+#define PYTHON_HTTP_HEADER_PLUGIN "python-http-header"
+
+struct _PythonHttpHeaderPlugin
+{
+  LogDriverPlugin super;
+
+  gchar *class;
+  GList *loaders;
+
+  GList *last_headers;
+  time_t last_run;
+  int timeout;
+
+  GHashTable *options;
+
+  struct
+  {
+    PyObject *class;
+    PyObject *instance;
+    PyObject *get_headers;
+  } py;
+};
+
+static gboolean
+_py_append_pylist_to_list(PyObject *py_list, GList **list)
+{
+  gchar *str;
+  PyObject *py_str;
+
+  if (!py_list || !PyList_Check(py_list))
+    goto exit;
+
+  Py_ssize_t len = PyList_Size(py_list);
+  for (int i = 0; i < len; i++)
+    {
+      py_str = PyList_GetItem(py_list, i); // Borrowed reference
+      if (!py_str || !PyUnicode_Check(py_str))
+        goto exit;
+
+      if (!(str = PyUnicode_AsUTF8(py_str)))
+        goto exit;
+
```

Comment:
This should probably use _py_get_string_as_string() which handles various python versions better. (e.g. not-unicode strings, which are the default on python2)